### ISSUE TITLE
--add-host: No connection via Domain. Connection via IP, only

### ISSUE THREAD

**Issue by Moltey** [containers/podman#12583]
/kind bug

**Description**

Using the **--add-host** parameter on a container does not work as expected on rootless podman. The same config is working on docker.

**Steps to reproduce the issue:**

1. Start mosquitto rootless on host with IP 192.168.0.11 (bound to sub.domain.com):
```bash
$ podman run -d \
  --name=mosquitto \
  --runtime crun \
  --restart on-failure \
  -p 8883:8883 \
  -p 8083:8083 \
  -v ~/mosquitto/mosquitto.conf:/mosquitto/config/mosquitto.conf \
  -v ~/mosquitto/data:/mosquitto/data \
  -v ~/mosquitto/log:/mosquitto/log \
  -v ~/certbot/1020/letsencrypt:/etc/letsencrypt:ro \
  --cap-drop=all \
  --cap-add=CAP_NET_BIND_SERVICE \
  --user 1020:1020 \
  eclipse-mosquitto:latest
```

2. Start node-red rootless on the same host with IP 192.168.0.11 (bound to sub.domain.com): 
``` bash 
$ podman run -d \
  --name=node-red \
  --restart on-failure \
  -v ~/node-red/data:/data \
  --user 1060:1060 \
  --add-host=sub.domain.com:191.168.0.11 \
  --net slirp4netns:allow_host_loopback=true \
  nodered/node-red:latest
```

**Describe the results you received:**

I am not able to connect mosquitto while using domain name sub.domain.com. Connection attempts to sub.domain.com are failing with a timeout:

``` bash
$ podman exec -it node-red curl -v mqtt://sub.domain.com:8883/
*   Trying 191.168.0.11:8883...
* connect to 191.168.0.11 port 8883 failed: Connection refused
* Failed to connect to sub.domain.com port 8883 after 76789 ms: Connection refused
* Closing connection 0
curl: (7) Failed to connect to sub.domain.com port 8883 after 76789 ms: Connection refused
```

**Describe the results you expected:**

However, connection attempts with the IP address are successful (at least a response from mosquitto):
```bash
$ podman exec -it node-red curl -v mqtt://192.168.0.11:8883/
*   Trying 192.168.0.11:8883...
* Connected to 192.168.0.11 (192.168.0.11) port 8883 (#0)
* Using client id 'curl4fb2f297'
> MQTT<
       curl4fb2f297mqtt_doing: state [0]
< mqtt_doing: state [2]
< Expected 0000 but got 0100
* Closing connection 0
curl: (8) Expected 0000 but got 0100
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
$ podman --version
podman version 3.2.1
```

**Output of `podman info --debug`:**

```
$ podman info --debug
host:
  arch: amd64
  buildahVersion: 1.21.3
  cgroupControllers: []
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpus: 4
  distribution:
    distribution: ubuntu
    version: "21.10"
  eventLogger: journald
  hostname: podman-host
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 5001
      size: 1
    - container_id: 1
      host_id: 500001
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 5001
      size: 1
    - container_id: 1
      host_id: 500001
      size: 65536
  kernel: 5.13.0-22-generic
  linkmode: dynamic
  memFree: 15415144448
  memTotal: 33395744768
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/5001/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.0.1
      commit: 6a7b16babc95b6a3056b33fb45b74a6f62262dd4
      libslirp: 4.4.0
  swapFree: 17179865088
  swapTotal: 17179865088
  uptime: 25h 39m 9.47s (Approximately 1.04 days)
registries:
  search:
  - docker.io
store:
  configFile: /home/podman/.config/containers/storage.conf
  containerStore:
    number: 6
    paused: 0
    running: 5
    stopped: 1
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/podman/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 9
  runRoot: /run/user/5001/containers
  volumePath: /home/podman/.local/share/containers/storage/volumes
version:
  APIVersion: 3.2.1
  Built: 0
  BuiltTime: Thu Jan  1 00:00:00 1970
  GitCommit: ""
  GoVersion: go1.16.7
  OsArch: linux/amd64
  Version: 3.2.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
$ apt list podman
Listing... Done
podman/impish,now 3.2.1+ds1-2ubuntu3 amd64 [installed,automatic]
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

No (I have to trust on Ubuntu updates), I checked the Troubleshooting Guide, though.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical

**Comment by Moltey** [comment]
The IP address 192.168.0.11 is the address of the physical NIC of the host. Therefore, mosquitto is bound e.g. to port 192.168.0.11:8883 of the host. I am able to connect to 192.168.0.11:8883 from other physical hosts in the network.

**Comment by Luap99** [comment]
Please check if the entry is added to the containers /etc/hosts file.

**Comment by Moltey** [comment]
Yes, it is:
``` bash
$ podman exec -it node-red cat /etc/hosts
127.0.0.1 localhost
127.0.1.1 podman-host

# The following lines are desirable for IPv6 capable hosts
::1     ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
191.168.0.11 sub.domain.com

10.88.3.1 host.containers.internal
```

The curl command initially provided also shows that the /etc/hosts entry is recognized correctly as curl tries to connect to 192.168.0.11 as it correctly resolves sub.domain.com to 192.168.0.11 ...

**Comment by Luap99** [comment]
OK that makes zero sense, what the hell is going on here?
If the curl connection directly to the ip:port works and curl successfully resolves the given name to the correct ip then it has to work. There should be no difference at the transport level. Can something in the mqtt protocol cause this?

**Comment by TomSweeneyRedHat** [comment]
Just a WAG, but if you add `--privileged` to your first command:

```
$ podman run -d \
 --privileged \
  --name=mosquitto \
...
```
Does that work?

**Comment by Moltey** [comment]
I am **_very, very, very sorry_** for wasting everyones time. I lost 5 hours of my life (and almost my mind) due to a typo:

What I had in my script:
--add-host=sub.domain.com:19`1`.168.0.11

What was ment to be:
--add-host=sub.domain.com:19`2`.168.0.11

After I tried the priviliged version of @TomSweeneyRedHat (not successfully), I tried to curl a new nginx conatiner on the same host as suggested by @Luap99 . The last one brought up the typo. Thank both of you!
